### ISSUE TITLE
CMake issues with MSYS2

### ISSUE THREAD

**Issue by hgdagon** [mgba-emu/mgba#1979]
## Warnings
A warning about `libavscale` missing, `libswscale` seemingly detected just fine.
And another one about `sqlite3` (capitalization issue)?

## Errors
Each target complains that `bzip2` and `lzma` were not found. No search is set up for either of the packages.

## Truncated Log

```cmake
----------
CMake Warning (dev) at D:/msys64/mingw64/share/cmake-3.19/Modules/FindPackageHandleStandardArgs.cmake:424 (message):
  The package name passed to `find_package_handle_standard_args` (SQLite3)
  does not match the name of the calling package (sqlite3).  This can lead to
  problems in calling code that expects `find_package` result variables
  (e.g., `_FOUND`) to follow a certain pattern.
Call Stack (most recent call first):
  D:/msys64/mingw64/share/cmake-3.19/Modules/FindSQLite3.cmake:54 (find_package_handle_standard_args)
  src/platform/cmake/FindFeature.cmake:12 (find_package)
  CMakeLists.txt:507 (find_feature)
This warning is for project developers.  Use -Wno-dev to suppress it.
----------
CMake Warning at src/platform/cmake/FindFeature.cmake:16 (message):
  Requested module libavresample missing for feature USE_LIBAVRESAMPLE.
  Feature disabled.
Call Stack (most recent call first):
  CMakeLists.txt:514 (find_feature)
-------------------
Configuring done
CMake Error at CMakeLists.txt:872 (add_library):
  Target "mgba" links to target "BZip2::BZip2" but the target was not found.
  Perhaps a find_package() call is missing for an IMPORTED target, or an
  ALIAS target is missing?


CMake Error at CMakeLists.txt:872 (add_library):
  Target "mgba" links to target "LibLZMA::LibLZMA" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?


CMake Error at src/platform/sdl/CMakeLists.txt:112 (add_executable):
  Target "mgba-sdl" links to target "BZip2::BZip2" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?


CMake Error at src/platform/sdl/CMakeLists.txt:112 (add_executable):
  Target "mgba-sdl" links to target "LibLZMA::LibLZMA" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?


CMake Error at src/platform/qt/CMakeLists.txt:289 (add_executable):
  Target "mgba-qt" links to target "BZip2::BZip2" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?


CMake Error at src/platform/qt/CMakeLists.txt:289 (add_executable):
  Target "mgba-qt" links to target "LibLZMA::LibLZMA" but the target was not
  found.  Perhaps a find_package() call is missing for an IMPORTED target, or
  an ALIAS target is missing?


Generating done
```

**Comment by Testsr** [comment]
Partial duplicate of #1857.

**Comment by endrift** [comment]
the libavscale/libswscale warnings are harmless. You usually won't have both.

The sqlite3 warning is frustrating but _mostly_ harmless.

The rest is covered by #1857

**Comment by hgdagon** [comment]
@endrift As harmless as they are, shouldn't they be suppressed? I'll try the `LIBZIP` workaround mentioned in the other issue.

**Comment by endrift** [comment]
The hard part is that you want one or the other. If both show up, it's not harmless. If one shows up, it is. It doesn't matter which one. That makes it somewhat more challenging.

**Comment by hgdagon** [comment]
@endrift But `libavresample` is deprecated anyway, right? So, most people are going to see this warning.

**Comment by endrift** [comment]
You're correct, but if you have libavresample and not libswresample you'll still see a warning, which is equally as harmless.

**Comment by hgdagon** [comment]
@endrift My point is, why don't we make this a dev warning, that users never see? Since, the way I understand it, both are there for compatibility across different systems.

**Comment by endrift** [comment]
I'd need to create a way to pass that information down through the dependency lookup. Either it's a dev warning for all of the feature detection or a dev warning for none of them right now. It's not a simple fix.

**Comment by hgdagon** [comment]
@endrift On an unrelated note, while the `LIBZIP` workaround did let me generate the project, the resulted executables aren't statically linked, even though I disabled `BUILD_SHARED`, enabled `BUILD_STATIC`, and made sure to point to static libs everywhere.

**Comment by endrift** [comment]
I don't think it's possible to make a static build with msys2. They just don't distribute all of the needed binaries as static, as far as I know.

**Comment by hgdagon** [comment]
@endrift I thought that would only be the case if we used `cmake .. -G "MSYS Makefiles"` straight, without any changes. And I used the GUI.
The only actual problem I see is `Qt5`, which is the everlasting MSYS2 problem. 
However, MSYS2 does distribute *ALL* the static libraries needed.
I would create another PR, addressing this, however, mGBA is targeting way too many platforms, than I can keep my eyes on. I might create more problems than I can solve.

**Comment by endrift** [comment]
Hm, I know there's a qt5-static package but I've never gotten it working when building.

**Comment by hgdagon** [comment]
If I took a shot at it, I would keep `Qt5` on user's consideration. Since a lot of MSYS2 users, myself included, are not willing to install that package. It's just not worth the hassle.

**Comment by endrift** [comment]
I'm not sure what you mean. You can't build mGBA without Qt and get a GUI.

**Comment by hgdagon** [comment]
I mean the user should have the choice of using `Qt5` or `Qt-static`, but all the other libraries can and should be linked statically, when `BUILD_STATIC` is selected.

**Comment by endrift** [comment]
Oh, yes, I agree. Are you saying they're not currently?

**Comment by hgdagon** [comment]
Nope...

`mgba-sdl.exe` I just built requires `zlib1.dll`, `libminizip-1.dll` and `SDL2.dll`, which, if you ask me is still 3 libs more than necessary.

The real party is over at `mgba.exe`, and all these guys are invited:

`avcodec-58.dll, avfilter-7.dll, avformat-58.dll, avutil-56.dll, libaom.dll, libass-9.dll, libbluray-2.dll, libbrotlicommon.dll, libbrotlidec.dll, libbz2-1.dll, libcelt0-2.dll, libcrypto-1_1-x64.dll, libdav1d.dll, libdouble-conversion.dll, libexpat-1.dll, libffi-7.dll, libfontconfig-1.dll, libfreetype-6.dll, libfribidi-0.dll, libgcc_s_seh-1.dll, libglapi.dll, libglib-2.0-0.dll, libgmp-10.dll, libgnutls-30.dll, libgraphite2.dll, libgsm.dll, libharfbuzz-0.dll, libhogweed-6.dll, libiconv-2.dll, libicudt67.dll, libicuin67.dll, libicuuc67.dll, libidn2-0.dll, libintl-8.dll, libLLVM.dll, liblzma-5.dll, libmfx-1.dll, libminizip-1.dll, libmodplug-1.dll, libmp3lame-0.dll, libnettle-8.dll, libogg-0.dll, libopencore-amrnb-0.dll, libopencore-amrwb-0.dll, libopenjp2-7.dll, libopus-0.dll, libp11-kit-0.dll, libpcre-1.dll, libpcre2-16-0.dll, libpng16-16.dll, librtmp-1.dll, libspeex-1.dll, libsqlite3-0.dll, libsrt.dll, libstdc++-6.dll, libtasn1-6.dll, libtheoradec-1.dll, libtheoraenc-1.dll, libunistring-2.dll, libvorbis-0.dll, libvorbisenc-2.dll, libvpx-1.dll, libvulkan-1.dll, libwavpack-1.dll, libwebp-7.dll, libwebpmux-3.dll, libwinpthread-1.dll, libx264-159.dll, libx265.dll, libxml2-2.dll, libzstd.dll, postproc-55.dll, Qt5Core.dll, Qt5Gui.dll, Qt5Multimedia.dll, Qt5Network.dll, Qt5Widgets.dll, qwindows.dll, SDL2.dll, swresample-3.dll, swscale-5.dll, xvidcore.dll, zlib1.dll`

Most of these belong to `ffmpeg`.

**Comment by hgdagon** [comment]
Also, `mgba-sdl.exe` has no icon and won't run at all... But, it's in the latest release, so I guess that's an unrelated issue.

**Comment by endrift** [comment]
you need to pass a ROM to it on the command line otherwise it'll just exit.

**Comment by hgdagon** [comment]
I see. Anyway, I'll experiment with static linking. I'll start with `ffmpeg`. Will comment back if get anywhere.

**Comment by endrift** [comment]
I was able to get 90% of the way there by using `CMAKE_LINK_SEARCH_START_STATIC` and merging the *_STATIC_LIBRARIES vars into the *_LIBRARIES vars in FindFeature.cmake, but then I ran into issues with the actual static packages:

- libxvidcore.a is called xvidcore.a for some reason, and this is not particularly easy to work around
- libvulkan.a doesn't exist at all it seems
- libcelt0.a and libopus.a have a bunch of overlapping symbols

None of these are really solvable on my end, and it's part of why I have custom builds of all of these packages that I distribute in the Docker images I use for building on Windows.

**Comment by hgdagon** [comment]
`libvulkan` is in a [separate package](https://packages.msys2.org/package/mingw-w64-x86_64-vulkan-loader?repo=mingw64), and, CONGRATULATIONS, we've found one of those few packages, that don't come with a static lib.

And, yeah... not a lot can be done on our side. Apparently, statically linking `ffmpeg` is another ongoing msys2 issue, that I wasn't aware of, along with `libcelt0` and `libopus`'s overlapping symbols.

**Comment by endrift** [comment]
Warnings have been fixed in 9397eee606ce7736131d83bfa5ae43919c737529; the errors are a dupe, so I'm closing this.
